libr/util/logging.c

```diff
@@ -0,0 +1,157 @@
+static char output_buf[LOGGING_OUTPUTBUF_SIZE] = ""; // Big buffer for building the output string
+
+/* TODO: Might want to create a r_logging_use_envvars function so that programs
+   that use r_util without core/cons have some way of changing the above vars
+   to the environmental variables. Also create seperate setter functions */
+
+// Used in cb_logging_config_level to check if input is integers only
+static inline bool str_is_digits(char *str, int len) {
+	int i;
+	for (i = 0; i < len; i++) {
+		if (!isdigit (str[i]))
+			return false;
+	}
+	return true;
+}
+
+// cconfig.c configuration callback functions below
+int cb_logging_config_level(void *coreptr, void *nodeptr) {
+	RConfigNode *node = (RConfigNode*)nodeptr;
+	char *value = node->value;
+	int value_len = strnlen (value, LOGGING_CONFIGSTR_SIZE);
+	if (!str_is_digits (value, value_len))
+		return false;
+	cfg_loglvl = atoi (value);
+	return true;
+}
+
+int cb_logging_config_traplevel(void *coreptr, void *nodeptr) {
+	RConfigNode *node = (RConfigNode*)nodeptr;
+	char *value = node->value;
+	int value_len = strnlen (value, LOGGING_CONFIGSTR_SIZE);
+	if (!str_is_digits (value, value_len))
+		return false;
+	cfg_logtraplvl = atoi (value);
+	return true;
+}
+
+int cb_logging_config_file(void *coreptr, void *nodeptr) {
+	RConfigNode *node = (RConfigNode*)nodeptr;
+	char *value = node->value;
+	int value_len = strnlen (value, LOGGING_CONFIGSTR_SIZE) + 1;
+	strncpy (&cfg_logfile, value, value_len);
+	return true;
+}
+
+int cb_logging_config_srcinfo(void *coreptr, void *nodeptr) {
+	RConfigNode *node = (RConfigNode*)nodeptr;
+	char *value = node->value;
+	switch (value[0]) {
+		case 't':
+		case 'T':
+			cfg_logsrcinfo = true;
+			break;
+		default:
+			cfg_logsrcinfo = false;
+	}
+	return true;
+}
+
+int cb_logging_config_colors(void *coreptr, void *nodeptr) {
+	RConfigNode *node = (RConfigNode*)nodeptr;
+	char *value = node->value;
+	switch (value[0]) {
+		case 't':
+		case 'T':
+			cfg_logcolors = true;
+			break;
+		default:
+			cfg_logcolors = false;
+	}
+	return true;
+}
+
+/**
+ * \brief Set the logging callback function for printing
+ * \param cbfunc RLoggingFuncdef style function to be called
+*/
+void r_logging_setcb(RLoggingFuncdef cbfunc) {
+	cb_logfunc_print = cbfunc;
+}
+
+/**
+ * \brief Internal logging function used by preprocessor macros
+ * \param funcname Contains the function name of the calling function
+ * \param filename Contains the filename that funcname is defined in
+ * \param lineno The line number that this log call is being made from in filename
+ * \param lvl Logging level for output
+ * \param tag String that will be prepended to the log line (describes lvl)
+ * \param fmtstr A printf like string
+ * This function is used by the R_LOG_* preprocessor macros for logging/debugging
+*/
+void _r_logging_internal(const char *funcname, const char *filename,\
+ unsigned int lineno, unsigned int level, const char *tag, const char *fmtstr, ...) {
```

Comment:
since `level` and `tag` are always the same, I see no reasons to have both arguments. Just compute the `tag` from the level, as we were doing in the previous log implementation.

Reply:
I wasn't sure if there would ever be a use-case where someone wanted to set a custom tag. That way they can throw something at the error level but set a custom tag like `GREPFORME` or whatever they want to kinda make it stand out. I guess they could always just have it in the string itself so it would be `ERROR: GREPFORME: text`